### DISCUSSION TITLE
Breaking changes in C# 10

### DISCUSSION THREAD

**Discussion by sakno** [dotnet/csharplang#5157] [Q&A]
C# 10 introduces lambda type inference which is breaking change in my case. My library has the following methods:
```csharp
void Method(Action action);
void Method(Expression expr);
```

Prior to C# 10 the consumer has no problem when invoking `Method(() => DoSomething())`. But now the delegate type must be explicitly defined: `Method((Action)() => DoSomething())`. Otherwise, the compilation process stopped with `error CS0121: The call is ambiguous`

**Comment by Joe4evr** [comment]
[This is already known to the language team.](https://github.com/dotnet/csharplang/blob/main/meetings/2021/LDM-2021-08-23.md#lambda-expression-and-method-group-type-inference-issues)

**Reply by sakno** [reply]
I see **Conclusion**: _No changes_ with the comment:
>  We think this is a fine workaround, especially since this API is already ambiguous in many cases, and will not be making any changes here.

But overloads mentioned before is not ambiguous in C# 9. Changing method names to avoid this ambiguity in API is breaking change. So my question why the language update makes correct programs incorrect?

**Reply by 333fred** [reply]
As Joe stated, we know this is going to cause some pain. In this case, you're experiencing a break because the lambda has a natural type, and is therefore convertible to Expression. Between Action and Expression, there is no best common type (because they are not convertible to each other), and thus you now get an ambiguous call.

This change is not going to be without some pain. It's one of the biggest, if not _the_ biggest, breaking changes we've ever made to the language. However, we think it's the correct change for the future of the language. IMO, an API like this is already conceptually ambiguous, even if overload resolution can pick between them because lambdas previously didn't have a natural type.

**Reply by sakno** [reply]
It's a big pain to me as a library author. I'm not going to change public API in incompatible manner. However, when the users come to me and ask _why their code cannot be compiled anymore_ what should I say?

**Reply by CyrusNajmabadi** [reply]
>  when the users come to me and ask why their code cannot be compiled anymore what should I say?

Tell them a source-breaking change happened in C# 10 and that we made the call that the pain here was acceptable due to the overall improvement to the language.

> It's a big pain to me as a library author.

Understood.  That's why we're wary to make breaks generally.  This met our bar though for overall value provided.

**Reply by sakno** [reply]
> Tell them a source-breaking change happened in C# 10

This is not realistic. Currently I have to rewrite even my tests for the library. This feature killing my DSL literally and forces to use `new DelegateType(() => { }` everywhere instead of plain `() => { }`.

If C# is community-driven language how I can vote/downvote this feature?

**Reply by jnm2** [reply]
@333fred @CyrusNajmabadi I don't understand. The linked meeting notes talk about ambiguity between parameter lists `(object, Action)` and `(int, Expression)`, where adding a new overload allows library authors to fix the problem, not between `(Action)` and `(Expression)` which seems to leave us with no fix.

**Reply by CyrusNajmabadi** [reply]
> If C# is community-driven language how I can vote/downvote this feature?

This discussion serves that purpose well.

**Reply by acaly** [reply]
> In this case, you're experiencing a break because the lambda has a natural type, and is therefore convertible to Expression. Between Action and Expression, there is no best common type (because they are not convertible to each other), and thus you now get an ambiguous call.

If lambda has a natural type, why that natural type (in this case, `Action`) is not having a higher priority in overload matching?

**Reply by 333fred** [reply]
> If lambda has a natural type, why that natural type (in this case, `Action`) is not having a higher priority in overload matching?

It's a natural type, but it's not an identity conversion. It's still a function type conversion, so it doesn't automatically win.

**Reply by 333fred** [reply]
Of note, we are still talking about how we want to resolve something like this, and all of the solutions we're currently talking about would address this case by causing the Action overload to be preferred (either by preferring all real delegate types over Expression, like they would be preferred over Delegate) or by causing natural type conversions to be less preferable (which means that the Expression overload would be less preferred than the Action one).

**Reply by TonyValenti** [reply]
@sakno - Something I've done in the past is rename my methods from:
```
void Method(Action action);
void Method(Expression expr);
```
to:
```
void MethodAction(Action action);
void MethodExpression(Expression expr);
```

I haven't had to use that to resolve a source-breaking change but to resolve situations where a method group has an item for consideration that cannot (and I think, should not) be considered because it contains types that are not accessible to the calling assembly.

**Reply by sakno** [reply]
@TonyValenti , I thought about that but its' breaking changes in API dictated by breaking changes in programming language.

**Reply by sakno** [reply]
@333fred , I think that `Expression` has lower probability of usage in public API rather than delegate type. So yes, the viable change to the compiler is to prefer delegate type over expression type.

**Reply by 333fred** [reply]
> So yes, the viable change to the compiler is to prefer delegate type over expression type.

It is one possible change, yes. As I said, we're considering multiple ways of addressing both this and a few other potential breaks from our lambda changes.

**Reply by sakno** [reply]
RC2 is here already. Is there any decision?

**Reply by 333fred** [reply]
https://github.com/dotnet/csharplang/blob/main/meetings/2021/LDM-2021-09-20.md#lambda-breaking-changes

**Reply by jnm2** [reply]
RC2 already shipped the fix described at that link, too: https://github.com/dotnet/core/issues/6570#issuecomment-940164245

**Reply by sakno** [reply]
In some circumstances Roslyn still tries to resolve overload in favor of `Expression` type:
```csharp
public sealed class Scope
{
    public void DoAction(Expression expr)
    {
    }
}

public static class ScopeExtensions
{
    public static void DoAction(this Scope scope, Action action)
    {
    }
}

var scope = new Scope();
scope.DoAction(() =>
{
    Console.WriteLine();
    Console.WriteLine();
});
```

This code cannot be compiled due to CS0834 `A lambda expression with a statement body cannot be converted to an expression tree`

**Reply by sakno** [reply]
However, the fix is working when both methods are instance or static methods. In LDM notes, there is a section `Method groups converting to Expression`. It states that
> Unlike the previous examples, however, we don't have any reports of this issue

Now this report exists.

**Reply by cston** [reply]
> In some circumstances Roslyn still tries to resolve overload in favor of `Expression` type:

Yes, for instance and extension methods, the compiler will use the containing type or first namespace scope that contains any applicable overloads. Since overloads with `Delegate` or `Expression` may be applicable in C#10, that is a breaking change. There is a similar issue with applicable methods on derived and base types. See item 2 in [Compiler Breaking Changes - DotNet 6.md](https://github.com/dotnet/roslyn/blob/main/docs/compilers/CSharp/Compiler%20Breaking%20Changes%20-%20DotNet%206.md).

**Reply by 333fred** [reply]
> However, the fix is working when both methods are instance or static methods. In LDM notes, there is a section `Method groups converting to Expression`. It states that
> 
> > Unlike the previous examples, however, we don't have any reports of this issue
> 
> Now this report exists.

Of note, that would not fix your scenario. That specifically covers _method groups_, ie `scope.DoAction(SomeMethod)`. This break is occurring because of our general rules around how extension methods are resolved: if an applicable instance method is found, extension methods aren't even searched. The `DoAction(Expression)` method is applicable, even if it has errors, so `DoAction(this Scope, Action)` is never even considered. 99% of the time this pattern occurs, it represents someone providing a strongly-typed delegate method that forwards to the weakly-typed instance method, so we considered it fine (and even desirable in many cases) to break.

**Reply by CyrusNajmabadi** [reply]
FWIW, i would like us to revisit: `The DoAction(Expression) method is applicable, even if it has errors,` in the future.  I think for lambdas it's a bit nonsensical.  We already understand for normal overload resolution "oh, this takes a string, but you passed an object, so that's an error, so don't consider this".  In think lambda+expression or lambda+delegatetype needs something similar.  We can see it, and try it out.  But if it's an error, then it should not count and we should keep looking.

I get that that's not what we do today.  But, IMO, that's fairly unintuitive and leads to these undesirable situations.  When you have instance+extension for me, the intuition *should be* "use the instance *if it works*, otherwise fallback and look at extensions".  It shouldn't be "oh, you found a non-functional instance?  well we're going to use that, despite us knowing it definitely does not work".  That just seems 'broken' to me :)

**Reply by sakno** [reply]
Let's do a small recap. What I should expect from .NET 6? Will this resolution between extension and instance method be solved? If no, I have to inform the users of my library that their code might be broken. If so, I need to be prepared for negative feedback. According to my experience, most users recognize that all negative impact caused by breaking changes in .NET must be eliminated by the library author somehow. In other words, when the choice _library vs BCL_ happens, the last one treated as ultimate source of truth.

> The DoAction(Expression) method is applicable, even if it has errors, in the future.

When to expect this future?

**Reply by CyrusNajmabadi** [reply]
> When to expect this future?

We'd need a proposal, a champion, a spec for how this would work, approval by ldm, and enough time in a future release cycle to do this work. Currently absolutely no steps have been taken here, so I can't even give any sort of estimate on when this could be expected

**Reply by CyrusNajmabadi** [reply]
> Will this resolution between extension and instance method be solved? 

It has been solved in that we are accepting and documenting the breaking change in behavior. 

Existing users of your library who do not recompile will continue having binary compat with your lib. To help users who do recompile, you can add an instance method to the type instead of having it be an extension method.

**Reply by CyrusNajmabadi** [reply]
> In other words, when the choice library vs BCL happens, the last one treated as ultimate source of truth.

Yes. We are never happy about breaking changes. However, in this case we think the net benefit is high enough for most users and libraries to make this worthwhile.  Omelettes and eggs and all that.
